P2P Preheat Proposal (#142)
goharbor/community: proposals/p2p_preheat_proposal.md

**These there open sourced projects have very obviously complementary advantages to each others and the joint solution will definitely expand the scenarios of image lifecycle management and improve the securities, reliabilities, and efficiencies.**

## Proposal

### Overview

Preheat aims to let the P2P providers pull the images that match the related criteria (defined in the preheating policy or manually selected) into the P2P cache (Dragonfly supernode cache) or P2P network (Kraken no-cache) in advance to make sure the required images have been existing in the P2P network before the large scale runtimes start to pull and then to reduce the latency of “hotting” the P2P network as the 1st copy has already been there.

Preheat is a lightweight and loose-couple integration solution between Harbor and kinds of P2P providers. No dependencies for each other when doing the deployments. All the connections are set up through configurations and services calls are directly relying on the preheat API spec that defines several restful APIs.

> #### Story

Reviewer: ```suggestion
#### Story

** Provider Management:**

* As a system user,  steven can add a new P2P preheat provider by providing the related require info like 'endpoint' and 'credential';
*  As a system user,  steven can manage(view/search/update/delete) the existing P2P preheat providers;
*  As a system user,  steven can set the selected preheat provider as the system default one;
*  As a system user,  steven can enable or disable the selected preheat provider;

**Preheating Policy Management:**
*  As a project admin, frank can create a new preheat policy by setting the related data like 'provider', 'name', 'filters', 'criteria' and 'trigger' under the specified project;
*  As a project admin, frank can duplicate the selected preheating policy to create a new one under the specified project;
*  As a project admin, frank can manage(view/update/delete/disable) the existing preheating policies under the specified project;

**Execute Preheating Policies:**
*  As a project admin, frank can specify `event-based(onpush/onscan/onlabel)` trigger way when creating the preheating policy under the specified project to let policy automatically execute when relevant events occurred;
*  As a project admin, frank can specify `schedule(CRON settings)` trigger way when creating the preheating policy under the specified project to let policy periodically execute following the settled CRON period;
*  As a project admin, frank can manually execute the selected preheating policy under the specified project;
*  As a project admin, frank can check all the execution history of the selected preheating policy under the specified project;
```

> #### Story

Author: Good ideat,thx.

- A status API for reporting the preheating progress stats

As there may be different implementations of the above API for different P2P providers, we introduce the drivers to handle the differences. Other components use these drivers that have a unified interface to talk to the P2P providers.

The controller is the primary component to handle the overall preheating workflows including trigger preheating process per the request and track the progress of the triggered preheating. It also responds to the requests related to P2P endpoint management and preheating execution events querying.

The P2P endpoint instances are managed by the P2P provider manager and the related data are persisted in the database.

The preheat tracker is responsible for maintaining and persisting the preheating progress and the final history records.

The policy evaluator is designed to provide P2P preheating policy management and evaluation functions. Once the related criteria defined in the policy are matched, for example, the specified label is applied to the images or the pre-defined schedule time is coming, the controller would be notified to start a preheating process immediately.

> 

Reviewer: @kofj 

Can we add a section to describe the main preheating process will be handled by a `preheat` job? No need to cover the detailed job logic as it has been covered in the sequence diagram shown below.

Author: OK, I'll add this section later.

#### Workflow

The diagram shwon below describes the workflow of P2P Preheat feature:
![primary workflow of preheating function](images/p2p/primary-workflow-of-preheating-function.jpg)

> 1. When a user adds a P2p provider instance to Harbor, Harbor will check status of the instance first, and if the instance is healthy(Return 200 HTTP status code), it will be added successfully.

Reviewer: Are those descriptions aligned with the updated diagram?

Author: Yes.

2. The preheating function can be triggered by manual, scheduled policy or push event with filter.
3. The Harbor system administrator can add a Preheating policy , specify the type of artifact that is needed, and use a filter to select specific artifact.
4. When Preheating is triggered manually, by an event or by scheduled policy, the Harbor notifies the P2P provider to perform preheat operation with some artifact info(Like token and artifact for pulling) via JobService; upon receipt of the notification, the provider responds to the Harbor in response to the HTTP 200 status code to accept the request and starts pulling the artifact to the P2P network. If the preheating resource has exist in the network, provider should return the HTTP 200 status code. If Harbor and provider communication times out, Harbor will tries to retry 3 times.
5. After the preheating is started, the Harbor will circularly request the P2P provider to inquire the preheating execution progress, and until the state of preheating executed by the P2p provider reaches the final, such as FINISHED or FAILED, and Harbor will updates the preheating status information to the event log.

#### Policy-based driven

> The core feature of preheat is that the functionality is policy-driven. We plan to support three different policies: manual, periodically scheduled, and event-based. A policy corresponds to a preheat provider instance as a target. We can use filter in the policy to specify which repositories and tags can be preheat; we can use criteria to restrict filtered artifacts to be triggered when scanned, signed, or labeled; and we use trigger to set the trigger timing of the policy to be manual, periodic, or event-based. Event-based triggers can be set to execute preheat when scanned, when push is complete, or when labeled.

Reviewer: I think we can simplify the words here:

P2P preheat is driven by policy. The policy contains the following main sections:

basic info:
  - name
  - description 

provider:
   - p2p preheat provider with pattern `vendor name - endpoint`

Filters:
   - repositories filter: filter repositories by their names (double star pattern)
   - tag filter: filter tags

Criteria:
    - Signature: decided if only signed images can be preheated
    - vulnerability: decide if the images matching the expected vulnerability situations can be preheated
    - labels: decide if images with the specified labels can be preheated

Trigger:
    - None: manual
    - Event-Based: when the specified events occurred, including on-push, on-scan and on-label
    - Schedule: periodical executing following the specified CRON

Author: Changed

**These there open sourced projects have very obviously complementary advantages to each others and the joint solution will definitely expand the scenarios of image lifecycle management and improve the securities, reliabilities, and efficiencies.**

## Proposal

### Overview

Preheat aims to let the P2P providers pull the images that match the related criteria (defined in the preheating policy or manually selected) into the P2P cache (Dragonfly supernode cache) or P2P network (Kraken no-cache) in advance to make sure the required images have been existing in the P2P network before the large scale runtimes start to pull and then to reduce the latency of “hotting” the P2P network as the 1st copy has already been there.

Preheat is a lightweight and loose-couple integration solution between Harbor and kinds of P2P providers. No dependencies for each other when doing the deployments. All the connections are set up through configurations and services calls are directly relying on the preheat API spec that defines several restful APIs.

> #### Story

Reviewer: ```suggestion
### Story
```

**Preheating Policy Management:**

- As a project admin, Frank can create a new preheat policy by setting the related data like 'provider', 'name', 'filters', 'criteria' and 'trigger' under the specified project;
- As a project admin, Frank can duplicate the selected preheating policy to create a new one under the specified project;
- As a project admin, Frank can manage(view/update/delete/disable) the existing preheating policies under the specified project;

**Execute Preheating Policies:**

- As a project admin, Frank can specify `event-based(ON-PUSH/ON-SCAN/ON-LABEL)` trigger way when creating the preheating policy under the specified project to let policy automatically execute when relevant events occurred;
- As a project admin, Frank can specify `schedule(CRON settings)` trigger way when creating the preheating policy under the specified project to let policy periodically execute following the settled CRON period;
- As a project admin, Frank can manually execute the selected preheating policy under the specified project;
- As a project admin, Frank can check all the execution history of the selected preheating policy under the specified project;

> #### Architecture

Reviewer: ```suggestion
### Architecture
```

- A status API for reporting the preheating progress stats

As there may be different implementations of the above API for different P2P providers, we introduce the drivers to handle the differences. Other components use these drivers that have a unified interface to talk to the P2P providers.

The controller is the primary component to handle the overall preheating workflows including trigger preheating process per the request and track the progress of the triggered preheating. It also responds to the requests related to P2P endpoint management and preheating execution events querying.

The P2P endpoint instances are managed by the P2P provider manager and the related data are persisted in the database.

The preheat tracker is responsible for maintaining and persisting the preheating progress and the final history records.

The policy evaluator is designed to provide P2P preheating policy management and evaluation functions. Once the related criteria defined in the policy are matched, for example, the specified label is applied to the images or the pre-defined schedule time is coming, the JobService would be notified to start a preheating process immediately.

> The JobService is designed to execute preheating process. When the JobService be notified to Start a preheating process, it will check the P2P Provider's health status first. And then, it will notifies the P2P provider to perform preheat operation.

Reviewer: The meanings of this paragraph are not accurate. 

The main logic of the preheating process is executed in a preheat job which will be submitted and tracked by the Task Manager. The preheating execution management will also be built on top of the capabilities provided by the Task Manager.

#### Workflow

The diagram shwon below describes the workflow of P2P Preheat feature:
![primary workflow of preheating function](images/p2p/primary-workflow-of-preheating-function.jpg)

1. When a system administrator user adds a P2p provider instance to Harbor, Harbor will check status of the instance first, and if the instance is healthy(Return 200 HTTP status code), it will be added successfully.
2. The Harbor project administrator can add a preheating policy under special project , specify the type of artifact that is needed, and use a filter to select specific artifact.

> 3. The preheating function can be triggered by manual, scheduled policy or push event (ON-PUSH/ON-SCAN/ON-LABEL). Scheduled policy can be update to set CRON schedule, it will submit a periodical job to the JobService. Event-based can be update to set event-based trigger and subscribe ON-PUSH/ON-SCAN/ON-LABEL events. Manual policy will be list by the project administrator to select and execute.

Reviewer: ```suggestion
3. The preheating function can be triggered by manual, scheduled policy or push event (ON-PUSH/ON-SCAN/ON-LABEL). The scheduled policy can be updated to set the CRON schedule, it will submit a periodical job to the JobService. Event-based can be updated to set event-based trigger and subscribe ON-PUSH/ON-SCAN/ON-LABEL events. Manual policy will be list by the project administrator to select and execute.
```

> 4. When preheating is triggered manually, by an event or by scheduled policy, the Harbor will create and run a preheat JobService task. When a task starts running, it will check health status of the P2P provider first, if Provider is healthy it should return HTTP 200 status code to the task, the task will notifies the P2P provider to perform preheat operation with some artifact info(Like token and artifact for pulling); upon receipt of the notification, the provider responds to the Harbor in response to the HTTP 200 status code to accept the request and starts pulling the artifact to the P2P network. If the preheating resource has exist in the network, provider should return the HTTP 200 status code. If Harbor and provider communication times out or Provider health check failed, the task will tries to retry 3 times until success or failed.

Reviewer: ```suggestion
4. When preheating is triggered manually, by an event or by scheduled policy, the Harbor will create and run a preheat JobService task. When a task starts running, it will check the health status of the P2P provider first, if Provider is healthy it should return HTTP 200 status code to the task, the task will notify the P2P provider to perform to preheat operation with some artifact info(Like token and artifact for pulling); upon receipt of the notification, the provider responds to the Harbor in response to the HTTP 200 status code to accept the request and starts pulling the artifact to the P2P network. If the preheating resource has existed in the network, the provider should return the HTTP 200 status code. If Harbor and provider communication times out or Provider health check failed, the task will try to retry 3 times until success or failure.
```

5. After the preheating is started, the JobService will circularly request the P2P provider to inquire the preheating execution progress, and until the state of preheating executed by the P2P provider reaches the final, such as FINISHED or FAILED, and Harbor will updates the preheating status information to the event log.

#### Policy-based driven

P2P preheat is driven by policy. The policy contains the following main sections:

1. basic info:

- name
- description

2. provider:

- p2p preheat provider with pattern vendor name
- endpoint

> 1. Filters:

Reviewer: ```suggestion
3. Filters:
```

- repositories filter: filter repositories by their names (double star pattern)
- tag filter: filter tags

4. Criteria:

- Signature: decided if only signed images can be preheated
- vulnerability: decide if the images matching the expected vulnerability situations can be preheated
- labels: decide if images with the specified labels can be preheated

5. Trigger:

- None: manual
- Event-Based: when the specified events occurred, including on-push, on-scan and on-label
- Schedule: periodical executing following the specified CRON

### Interface and models

> #### Preheating Driver interface

Reviewer: As we'll leverage the Task Manager to submit and track the preheating jobs as well as to manage the execution history, some of the models related to the preheating execution management may need to be adjusted and modified.

	CheckProgress(taskID string) (*PreheatingStatus, error)
}
```

#### Structs job service

```go
type PreheatingStatus struct {
	TaskID     string `json:"task_id"`
	Status     string `json:"status"`
	Error      string `json:"error,omitempty"`
	StartTime  string `json:"start_time"`
	FinishTime string `json:"finish_time"`
}

// Jobs for preheat
// PreheatJob : the name of the P2P preheat job in job service
const PreheatJob = "PREHEAT_JOB"
// PREHEAT_SCHEDULER_JOB : the name of the P2P preheat scheduler job in job service
const PreheatSchedulerJob = "PREHEAT_SCHEDULER_JOB"

// Artifact type
const PreheatingArtifactTypeImage = "Image"

// Status of preheating job
const PreheatingStatusPending = "Pending"
const PreheatingStatusRunning = "Running"
const PreheatingStatusFinished = "Finished"
const PreheatingStatusFailed = "Failed"

// Preheat implements the job interface
type Preheat struct{}

// Scheduler is a job running in Jobservice which can be used as
// a scheduler when submitting it as a scheduled job. It receives
// a URL and data, and post the data to the URL when it is running
type Scheduler struct {}
```

#### Preheat Metadata Model

```go

> type Metadata struct {

Reviewer: Metadata = Provider? //P2P Preheat Provider?

	ID             int64  `orm:"pk;auto;column(id)" json:"id"`
	Name           string `orm:"column(name)" json:"name"`
	Description    string `orm:"column(description)" json:"description"`
	Provider       string `orm:"column(provider)" json:"provider"`
	Endpoint       string `orm:"column(endpoint)" json:"endpoint"`
	AuthMode       string `orm:"column(auth_mode)" json:"auth_mode"`
	AuthData       string `orm:"column(auth_data)" json:"auth_data"`
	Extensions     string `orm:"column(extensions)" json:"extensions"`
	Status         string `orm:"column(status)" json:"status"`
	Enabled        bool   `orm:"column(enabled)" json:"enabled"`
	SetupTimestamp int64  `orm:"column(setup_timestamp)" json:"setup_timestamp"`
}
```

#### Policy Model

```go
const EventTypeScanned = "Scanned"
const EventTypeSigned = "Signed"
const EventTypeLabeled = "Labeled"

type Trigger struct {
	// Type: manual, schedule, event
	Type string `json:"type"` // The preheat policy trigger type. The valid values ar manual, scheduled.
	// Event type: scanned, push, labeled

> 	Event          string `json:"event,omitempty"`

Reviewer: If it is event-based, I don't think we need to specify the concrete events. Any events occurred we need to do the similar process.

events -> look fro the matched policies -> evaluated -> tasks

> 	TriggerSetting struct {

Reviewer: `TriggerSetting` can be referred to with a pointer as it's possible to be optional.

		Cron string `json:"cron,omitempty"` // The cron string for scheduled trigger.
	} `json:"trigger_setting,omitempty"`
}

// FilterType represents the type info of the filter.
type FilterType string

// Filter holds the info of the filter
type Filter struct {
	Type  FilterType  `json:"type"`
	Value interface{} `json:"value"`
}

const CriteriaTypeScanned = "Scanned"
const CriteriaTypeSigned = "Signed"
const CriteriaTypeLabeled = "Labeled"

type PreheatPolicy struct {
	ID          int64     `orm:"column(id)" json:"id"`
	Name        string    `orm:"column(name)" json:"name"`
	Description string    `orm:"column(description)" json:"description"`
	ProjectID   int64     `orm:"column(project_id)" json:"project_id"`

> 	TargetID    int64     `orm:"column(target_id)" json:"target_id"`

Reviewer: ```suggestion
	ProviderID    int64     `orm:"column(providerID)" json:"target_id"`
	Provider       *Provider `orm:"column(-)" json:"provider"`
```

For providing full info to the client such as UI?

	Filters     []*Filter `orm:"column(-)" json:"-"`

> 	FiltersStr  string    `orm:"column(filters)" json:"filters"`

Reviewer: ```suggestion
	FiltersStr  string    `orm:"column(filters)" json:"-"`
```
?

	// Criteria type: scanned、signed、labeled

> 	Criteria    string    `orm:"column(criteria)" json:"criteria"`

Reviewer: +1

> 	Criteria    string    `orm:"column(criteria)" json:"criteria"`

Author: Or change it to below.
```go
       // For DAO
	CriteriaStr    string    `orm:"column(criteria)" json:"-"`
      // For others
	Criteria       []string    `orm:"column(-)" json:"criteria"`
```

	Trigger     Trigger   `orm:"column(-)" json:"-"`

> 	TriggerStr  string    `orm:"column(trigger)" json:"trigger"`

Reviewer: ```suggestion
	TriggerStr  string    `orm:"column(trigger)" json:"-"`
```
?

> 	Enabled     bool      `orm:"column(enabled)" json:"enabled"`

Reviewer: ```suggestion
	Enabled     bool      `orm:"column(enabled)" json:"enabled"`
	Default      bool       `orm:"column(isDefault)" json:"default"`
```

	CreatedAt   time.Time `orm:"column(creation_time)" json:"creation_time"`
	UpdatedTime time.Time `orm:"column(update_time)" json:"update_time"`
}
```

#### Event Model

```go

> type Execution struct {

Reviewer: ```suggestion

```
// REMOVE?

	ID          int64     `orm:"column(id)" json:"id"`
	PolicyID    int64     `orm:"column(policy_id)" json:"policy_id"`
	TriggerType string    `orm:"column(trigger_type)" json:"trigger_type"`
	TaskID      string    `orm:"column(task_id)" json:"task_id"`
	Artifacts   string    `orm:"column(artifacts)" json:"artifacts"`
	Status      string    `orm:"column(status)" json:"status"`
	Error       string    `orm:"column(result)" json:"result,omitempty"`
	StartTime   time.Time `orm:"column(start_time)" json:"start_time"`
	FinishTime  time.Time `orm:"column(finish_time)" json:"finish_time"`
}
```

### APIs

> #### 1. APIs for instance

Reviewer: Can we change `instance` to `provider` and "provider" property to `vendor`?

Create prepheat instance

```
POST /preheat/instances
Data:
{
    "name": "someInstance",
    "description": "this is draganfly instance.",
    "provider": "drangfly",
    "endpoint": "http://abc.com/",
    "auth_mode": "none",
    "enabled": false
}
```

List instances

```
GET /preheat/instances
```

Update instance

```

> PUT /preheat/instances/{id}

Reviewer: Set as default will only change one property. I'm wondering if the update API can support this use case?

Data:
{
    "description": "this is draganfly instance.",
    "provider": "drangfly",
    "endpoint": "http://abc.com/",
    "auth_mode": "none",
    "enabled": false
}
```

Delete instances

```
DELETE /preheat/instances/{id}
```

> ### 2. APIs for policy

Reviewer: As the policy is managed at the project level, so all the API endpoint should be changed to be the following pattern for consistence purpose:

`(/api/v2.0/projects/{project_name}/preheat/policies/{policy_id})`;

for executions bound with the given policy, the related API pattern can be 

`(/api/v2.0/projects/{project_name}/preheat/policies/{policy_id})/executions/{execution_id}`

Create Preheat policy

```
POST /preheat/policies
Data:
{
    "name": "by-cron",
    "description": "preheating by push cron",
    "target": 1,
    "trigger": {
        "type": "cron",
        "trigger_setting": {
            "cron": "*/1 * * * *"
        }
    },
    "filters": "",

>     "artifact_type": "image",

Reviewer: add `criteria` here for a clear description

    "enabled": true
}
```

List preheat policy

```
GET /preheat/policies
```

Update preheat policy
Data:
{
"name": "by-cron",
"description": "preheating by push cron",

> "target": 1,

Reviewer: ```suggestion
"provider": {
   id: 1,
   vendor: "dragonfly",
   endpoint: "https://dragonfly.com"
},
```
